File: raceEngine/src/engine/world.js
// La fabrique d'un monde : grille de depart, karts, decor, circuit.
// Tout ce qu'une course a besoin de savoir avant son premier pas.

import { randomRange, shuffleArray } from './math.js';
import { parkPosition } from './geometry.js';
import { deriveCharacterStats, fastestBoostedSpeed, getNewMomentumTarget } from './stats.js';
import { laneY } from './driving.js';
import { shadowCount, shadowFrom, shadowHi, shadowLo, shadowTo } from './vision.js';
import { countdownDuration } from './race.js';

// `startOrder` est l'ordre d'arrivee de la course precedente : le vainqueur
// repart en pole. Absent, la grille est tiree au sort. `grandPrix` reporte le
// bloc en cours ({ round, points }) ; absent, la course ouvre un bloc neuf.
function createWorldState(cfg, rng, now, startOrder, grandPrix) {
    const roadHeight = cfg.road.maxY - cfg.road.minY;
    const race = cfg.race;
    const countdownMs = countdownDuration(race);

    // Le circuit vient du dessin, pose sur la config par
    // raceEngine/src/track.js. Sans lui il n'y a pas de monde a construire :
    // autant le dire ici plutot que de faire tourner une course sur un tour
    // de largeur NaN, ou personne ne croise jamais de boite.
    const drawnBoxes = cfg.world.itemBoxes;
    if (!cfg.world.width || !drawnBoxes || !drawnBoxes.length) {
        throw new Error('cfg.world sans circuit : appeler track.applyTrack(cfg, circuit) '
            + 'avant createWorldState. Les circuits se dessinent dans tracks/.');
    }

    const itemBoxes = drawnBoxes.map(box => ({
        worldX: box.x,
        y: box.y,
        active: true,
        reactivateTime: 0
    }));

    // Les pipes ne connaissent aucun etat : ni actifs, ni repris, ni
    // detruits. Ils sont recopies ici quand meme, pour que tout le contenu
    // du monde se lise au meme endroit que le reste — et parce qu'un jour
    // l'un d'eux voudra peut-etre bouger.
    const pipes = (cfg.world.pipes || []).map(pipe => ({
        worldX: pipe.x,
        y: pipe.y,
        // Sa couleur, et elle s'arrete la : le moteur ne la lit nulle part.
        // Elle est recopiee pour la meme raison que le reste — tout le
        // contenu du monde se lit au meme endroit, et le decor se dessine
        // depuis l'etat, jamais depuis le tracé.
        kind: pipe.kind || 'green'
    }));

    const statsTable = deriveCharacterStats(cfg);
    const roster = Object.keys(statsTable);
    const names = (startOrder && startOrder.length === roster.length)
        ? startOrder.slice()
        : shuffleArray(roster, rng);

    const karts = [];
    const kartsById = {};

    names.forEach((charName, index) => {
        const row = Math.floor(index / race.grid.lanes.length);
        const col = index % race.grid.lanes.length;

        const gapToLine = race.grid.backOffset + row * race.grid.rowGap + col * race.grid.colStagger;
        let worldX = cfg.world.finishLineX - gapToLine;
        if (worldX < 0) worldX += cfg.world.width;

        const depth = race.grid.lanes[col] + row * race.grid.laneSlope;
        const verticalPos = Math.min(cfg.road.maxY,
                                     Math.max(cfg.road.minY, cfg.road.minY + roadHeight * depth));
        const stats = statsTable[charName];

        const kart = {
            id: index,
            charName: charName,
            worldX: worldX,
            yPercent: verticalPos,
            totalDistance: 0,

            stats: stats,
            absoluteVelocity: 0,
            momentum: 0,
            momentumTarget: getNewMomentumTarget(rng, cfg, stats),
            nextMomentumChange: now + randomRange(rng, cfg.speeds.momentumDriftMin, cfg.speeds.momentumDriftMax),

            // Elan mis de cote pendant un objet de vitesse, et ce qu'il
            // restait a son compte a rebours. -1 veut dire « rien en
            // attente » : c'est le seul etat qui autorise la mise de cote,
            // et le seul que produit un incident, qui refait l'elan et n'a
            // donc rien a rendre.
            preBoostMomentum: -1,
            preBoostDriftLeft: 0,
            vy: 0,
            targetVy: 0,

            // Canaux de choc, tenus a l'ecart du pilotage et du moteur.
            // `bumpVy` est en profondeur/s, `bumpVx` en pixels/s. Les deux
            // s'ajoutent au deplacement puis s'amortissent seuls : ecrire un
            // choc dans `vy` le faisait effacer par le volant avant que
            // les karts se soient decolles.
            bumpVy: 0,
            bumpVx: 0,

            // Vitesse le long de la piste sur le tick ecoule, en pixels/s.
            // Relevee a la fin du deplacement et lue par la passe de contact,
            // qui a besoin d'une vitesse de rapprochement reelle — recul de
            // pipe et tete-a-queue compris — et non de la consigne moteur.
            contactSpeed: 0,

            state: 'grid',
            rank: index + 1,

            aiState: 'cruising',

            hitEndTime: 0,

            // Choc contre un pipe. `bumpEndTime` porte l'arret net,
            // `bumpRecoilLeft` ce qu'il reste a reculer. Le sursis est
            // retenu par tuyau : un kart qui vient d'en heurter un doit
            // pouvoir se cogner au suivant.
            bumpEndTime: 0,
            bumpRecoilLeft: 0,
            bumped: false,

            // Ecrase par un kart reste grand. Date et drapeau, comme le
            // reste. La date ne depasse jamais celle du rapetissement : un
            // kart redevenu grand n'est plus plat.
            flatEndTime: 0,
            isFlat: false,

            // Contact en cours avec un tuyau pendant un tete-a-queue : la
            // toupie est bloquee tant qu'il tient, sans que ce soit un choc.
            pipeBlocked: false,
            pipeImmuneUntil: 0,
            lastPipeIndex: -1,

            // Tuyau en cours de contournement, et couloir choisi pour le
            // passer. L'index tient jusqu'a ce que le tuyau soit derriere :
            // c'est ce qui donne une trajectoire au lieu d'une suite
            // d'ecarts.
            pipeTargetIndex: -1,
            pipeLaneY: 0,

            heldItem: null,
            throwTime: 0,
            pendingItemGrantTime: 0,

            boostEndTime: 0,
            starEndTime: 0,
            isInvincible: false,
            hitInvincibleUntil: 0,

            // Rapetissement par l'eclair. La date pilote la simulation, le
            // booleen part dans le snapshot.
            shrinkEndTime: 0,
            isShrunk: false,

            // Bill Ball. `billAhead` retient qui reste a doubler : la vider
            // est ce qui raccourcit le vol.
            isBill: false,
            billStartedAt: 0,
            billEndTime: 0,
            billSlowUntil: 0,
            billAhead: [],

            trailTime: 0,
            brakeUntil: 0,

            // Quand il a traverse une zone de boxes — le coup d'oeil
            // arriere en depend (cf. `vision.boxGlanceMs`) — et l'episode de
            // danger pour lequel il a deja tranche entre bouclier et tir.
            boxPassedAt: -Infinity,
            shieldAt: -Infinity,
            shieldHold: false,

            // Severite du frein en cours, et prochaine occasion de decider
            // de se laisser doubler (cf. `vision.giveWay`).
            brakeFactor: 0,
            giveWayRetryAt: 0,
            shotDirection: 1,
            // Le plan de tir a-t-il ete fait en tete ? Il ne vaut plus rien
            // si le kart s'est fait doubler depuis.
            shotAsLeader: false,
            lobbing: false,
            aimError: 0,

            // Le releve d'un tir vers l'arriere : la profondeur vue lors du
            // coup d'oeil, et sa date. Il se perime (`vision.aimMemoryMs`),
            // et c'est voulu — viser de memoire, c'est viser ou l'autre
            // ETAIT.
            aimTargetY: 0,
            aimTargetAt: -Infinity,

            // Tout ce que le kart a percu au dernier balayage, et rien d'autre :
            // le pilotage ne lit plus le monde. Les deux dates de depart sont
            // decalees kart par kart pour qu'ils ne balayent ni ne tournent la
            // tete tous ensemble.
            sight: {
                at: now - cfg.vision.scanIntervalMs
                    + Math.round(index * cfg.vision.scanIntervalMs / names.length),
                back: false,
                backUntil: 0,

                // Sens du dernier balayage effectue, a distinguer de `back`
                // qui est l'attention du moment (cf. `perceive`).
                scanBack: false,

                nextGlance: now
                    + Math.round(index * cfg.vision.glanceIntervalMs / names.length),

                // Date du dernier coup d'oeil en arriere. C'est elle qui
                // autorise a viser derriere : avoir regarde, et non regarder
                // pendant. Loin dans le passe au depart — personne n'a encore
                // rien vu.
                seenKartY: 0,
                seenKartDist: -1,

                threatId: 0,
                threatKind: '',
                threatY: 0,
                threatTtc: Infinity,
                planGone: false,

                spans: [],
                spanCount: 0,

                // Ce que le balayage a vu passer sans le voir : ce qui
                // tombait dans l'ombre d'un corps plus proche. Purement
                // observable — aucune decision ne le lit. Cf. `perceive`.
                hiddenIds: [],
                hiddenCount: 0,

                // Les ombres elles-memes, telles que la marche les a
                // empilees : deux pentes depuis l'oeil, et les deux
                // distances entre lesquelles elles portent. Observable
                // aussi — la decision les a deja consommees.
                shadowLo: [],
                shadowHi: [],
                shadowFrom: [],
                shadowTo: [],
                shadowCount: 0,

                // Ou etait la camera pendant ce balayage. Sans elle, les
                // pentes ci-dessus ne se rattachent a rien.
                eyeBack: 0,
                eyeY: 0,

                // La portee du balayage courant, avant ou arriere. Elle se
                // deduit du sens et de `vision.range`, mais l'observateur
                // n'a pas a refaire ce choix : ce qui s'affiche doit etre ce
                // que le moteur a regarde.
                scanRange: 0,

                // Les profondeurs des karts qui roulent avec lui — cf.
                // `vision.crowd` et l'encombrement dans `laneRisk`.
                crowdY: [],
                crowdCount: 0,

                pipeIndex: -1,
                pipeDist: 0,
                pipeAheadIndex: -1,
                pipeAheadDist: 0,
                aheadKartY: 0,
                aheadKartDist: -1,
                boxY: 0,
                boxDist: -1,
                pressure: false,
                pressureY: 0,
                pressureId: 0,
                // De quel cote vient le releve ci-dessus : un porteur dans
                // le dos qui peut tirer, ou un porteur devant qui peut
                // lacher. Meme perception, deux decisions.
                pressureBack: false,

                // Le porteur qu'on SUIT, et depuis quand. Meme souvenir
                // date que `dangerAt` plus bas, meme peremption, pour la
                // meme raison : un coup d'oeil arriere ne doit pas effacer
                // ce que le kart a devant lui. Seul un balayage AVANT pose
                // ou leve ce souvenir.
                frontAt: -Infinity,
                frontY: 0,
                frontId: 0,

                // Le danger APERCU DERRIERE, et depuis quand. C'est un souvenir
                // et non un etat : le tirage du coup d'oeil n'a jamais lieu
                // pendant un coup d'oeil, donc il ne peut lire que ce qu'on a vu.
                // Sans lui, le kart oublierait la carapace entre deux
                // clignements.
                //
                // `dangerAt` est rafraichi a chaque coup d'oeil qui le revoit ;
                // `dangerSince` marque le debut de l'EPISODE, ce qui evite de
                // rejouer le choix du bouclier.
                //
                //   'shot'    une carapace en vol, deja lancee
                //   'carrier' quelqu'un derriere qui en porte une
                //   'ram'     une etoile ou un bill : rien a lui opposer
                dangerAt: -Infinity,
                dangerSince: -Infinity,
                dangerKind: '',

                // Les rouges apercues derriere : la plus proche, et combien.
                // Cf. `vision.giveWay`.
                redBehindDist: -1,
                redBehindY: 0,
                redBehindId: 0,
                redBehindCount: 0,

                // Et leur souvenir, sans lequel le releve ci-dessus
                // n'existait que pendant le coup d'oeil. Meme peremption
                // que `dangerAt`, et pose par le seul balayage arriere.
                redMemAt: -Infinity,
                redMemDist: -1,
                redMemY: 0,
                redMemId: 0,
                redMemCount: 0
            },

            // Le plan d'evitement en cours. Il survit a la perte de vue :
            // seule son echeance, ou le constat que la menace est passee,
            // le ferme. Cf. `updatePlan`.
            plan: {
                kind: '',
                threatId: 0,
                threatY: 0,
                laneY: verticalPos,
                dir: 0,
                intensity: 30,
                until: 0,
                reviewAt: 0,
                coarse: false,
                idle: false,
                stuck: false,
                crossing: false
            },

            // Menaces deja jugees : reflexe tire et verdict d'inattention,
            // retenus le temps qu'elles passent. Ecrase le premier
            // emplacement libre ou perime, sinon le plus ancien — cf.
            // `vision.memorySlots` et `vision.memoryMs`. Zero ne designe
            // aucune menace : les objets s'identifient a partir de 1, les
            // karts en negatif.
            judgedId: new Array(cfg.vision.memorySlots).fill(0),
            judgedSeenAt: new Array(cfg.vision.memorySlots).fill(-Infinity),
            judgedReactAt: new Array(cfg.vision.memorySlots).fill(0),
            judgedIgnored: new Array(cfg.vision.memorySlots).fill(false),

            // Prochaine chance de prendre une decision de securite, UNE PAR
            // COTE. Ce qui se retente est la decision et non la perception
            // de celui qui la provoque — mais se ranger devant un porteur
            // et se ranger derriere un porteur sont deux decisions, prises
            // sur deux dangers. Un compteur commun les faisait s'annuler
            // l'une l'autre (cf. `updatePlan`).
            safetyRetryFrontAt: 0,
            safetyRetryBackAt: 0,
            // Prochaine reprise du couloir de tuyau. Zero : le premier
            // couloir choisi est aussitot revisable (cf. steerAroundPipes).
            pipeReviewAt: 0,

            nextWanderTime: now + randomRange(rng, 1000, 5000),
            wanderEndTime: 0,
            wanderY: 0,

            // Gain de volant sous objet de vitesse. Neutre au depart, repose
            // a chaque tick (cf. `stepPhysics`).
            steerBoost: 1,

            // Distance perdue a la contrainte de virage depuis le depart, en
            // pixels de monde. Compteur d'observation, jamais lu par le jeu.
            cornerLostPx: 0,

            lapCount: 0,
            hasPassedFinishLine: false,
            stopped: false,

            // Cinq tours pleins, plus le bout de piste qui separe la place
            // de grille de la ligne. Ce segment initial ne compte pas comme
            // un tour : il vaut moins d'une seconde.
            finishDistance: race.laps * cfg.world.width + gapToLine,
            finished: false,
            finishRank: 0,
            startStallUntil: 0,

            currentSpinFrame: 0,

            // Dernier objet recu, lu par le tirage suivant pour freiner
            // deux fois de suite le meme.
            lastItem: null
        };

        karts.push(kart);
        kartsById[index] = kart;
    });

    return {
        cameraX: parkPosition(cfg, race.parkStartOffset),
        bgCameraX: 0,
        karts: karts,
        kartsById: kartsById,
        items: [],
        itemBoxes: itemBoxes,
        pipes: pipes,
        cachedLeader: null,
        rankedCount: 0,

        phase: 'countdown',
        countdownMs: countdownMs,
        startAt: now + countdownMs,
        resultsAt: 0,
        leaderLap: 1,
        finalSignShown: false,
        flagShown: false,
        finishOrder: [],

        // Grand prix. `gpRound` est le numero de cette course dans le bloc,
        // `gpPoints` le cumul par personnage a l'entree, `racePoints` ce que
        // cette course rapporte — vide tant qu'elle n'est pas close.
        gpRound: (grandPrix && grandPrix.round) || 1,
        gpPoints: Object.assign({}, (grandPrix && grandPrix.points) || null),
        racePoints: {},
        cameraSpeed: cfg.speeds.roadPPS,
        // Panneau tenu par Lakitu : { group, frame, until }.
        sign: { group: 'start', frame: 1, until: now + countdownMs + race.goSignMs },

        nextItemId: 1,
        // Decotes en cours, par type d'objet : absent vaut 1, c'est-a-dire
        // intact. Un type n'y entre qu'une fois sorti au moins une fois.
        itemDecay: {},
        // Garde-fou de delai propre a la bleue, arme des le depart pour ne
        // rien bloquer au premier tour.
        blueShellLastAt: now - cfg.blueShell.cooldownMs,
        // Plancher de vitesse du bill : ne depend que de la config.
        billFloorSpeed: fastestBoostedSpeed(cfg) * cfg.bill.minLeadRatio,
        // Orage en cours, ou null. Un seul a la fois.
        storm: null,
        previousRanking: [],
        lastLeaderboardUpdate: 0
    };
}

export {
    createWorldState,
};

File: raceEngineCpp/src/engine/world.cpp
// <- raceEngine/src/engine/world.js + stats.js
//
// La fabrique d'un monde. Deux choses : deduire les stats de chaque personnage,
// puis poser la grille de depart.

#include "engine/world.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>

#include "engine/math.hpp"

namespace engine {

StatsTable derive_character_stats(const config::Config& cfg) {
    const config::KartStatsCfg& spec = cfg.kartStats;
    const double span = (spec.maxPoints - spec.minPoints) != 0
        ? static_cast<double>(spec.maxPoints - spec.minPoints)
        : 1.0;

    StatsTable table;
    table.reserve(spec.characters.size());

    for (const config::CharacterSpec& raw : spec.characters) {
        // Un axe hors bornes ou un budget qui ne tombe pas juste est une erreur
        // d'AUTEUR : le message dit quoi corriger, la pile d'appels ne dirait
        // rien de plus.
        const int axes[3] = { raw.weight, raw.power, raw.handling };
        for (int v : axes) {
            if (v < spec.minPoints || v > spec.maxPoints) {
                throw std::runtime_error("kartStats : " + raw.name + " a un axe a "
                    + std::to_string(v) + ", hors de ["
                    + std::to_string(spec.minPoints) + ", "
                    + std::to_string(spec.maxPoints) + "]");
            }
        }
        const int total = raw.weight + raw.power + raw.handling;
        if (total != spec.budget) {
            throw std::runtime_error("kartStats : " + raw.name + " totalise "
                + std::to_string(total) + " points au lieu des "
                + std::to_string(spec.budget) + " du budget");
        }

        CharacterStats s;
        s.name = raw.name;
        s.rawWeight = raw.weight;
        s.rawPower = raw.power;
        s.rawHandling = raw.handling;

        s.normWeight   = (raw.weight - spec.minPoints) / span;
        s.normPower    = (raw.power - spec.minPoints) / span;
        s.normHandling = (raw.handling - spec.minPoints) / span;

        s.mass  = lerp(spec.mass.min, spec.mass.max, s.normWeight);
        s.force = lerp(spec.force.min, spec.force.max, s.normPower);
        // L'axe handling est courbe, pas droit : cf. `gripCurve` en config.
        s.grip  = lerp(spec.grip.min, spec.grip.max,
                       std::pow(s.normHandling, spec.gripCurve));

        // Pointe ADDITIVE : chaque axe apporte ses px/s, et les apporte seul.
        s.topSpeed = spec.speedBase
            + spec.speedPerWeight * s.normWeight
            + spec.speedPerPower * s.normPower;

        s.acceleration = clamp(s.force / std::pow(s.mass, spec.massDragAccel),
                               spec.accelClamp.min, spec.accelClamp.max);
        s.agility = clamp(s.grip / std::pow(s.mass, spec.massDragAgility),
                          spec.agilityClamp.min, spec.agilityClamp.max);

        // Ce que braquer coute en vitesse. Batie sur `agility`, elle laissait
        // `massDragAgility` gouverner deux choses a la fois : la vitesse a
        // laquelle un lourd tourne, ET ce que tourner lui coute.
        s.cornering = std::pow(s.grip, spec.cornerGripGain)
            * std::pow(s.force, spec.cornerPowerGain)
            / std::pow(s.mass, spec.cornerMassDrag);

        table.push_back(std::move(s));
    }

    return table;
}

WorldState create_world_state(const config::Config& cfg, Rng& rng, double now,
                              const std::vector<std::string>& startOrder) {
    WorldState state;
    state.simTime = now;
    state.statsTable = derive_character_stats(cfg);

    if (state.statsTable.empty()) {
        throw std::runtime_error("createWorldState : aucun personnage en config");
    }

    // L'ordre de la grille. Le roster de base suit l'ordre de la config, sans
    // les personnages retires du tirage (`enabled`, miroir de `roster.enabled`
    // du JS) ; on le reordonne selon `startOrder` quand une manche precedente
    // l'a decide, sinon on le melange — c'est ce qui ouvre un grand prix. Les
    // `kartCount` premiers courent : melange, c'est le tirage ; reordonne, ce
    // sont les karts de la manche precedente.
    std::vector<int> roster;
    roster.reserve(state.statsTable.size());
    for (size_t i = 0; i < state.statsTable.size(); i++) {
        if (cfg.kartStats.characters[i].enabled) roster.push_back(static_cast<int>(i));
    }
    if (roster.empty()) {
        throw std::runtime_error("createWorldState : aucun personnage actif (enabled)");
    }

    if (!startOrder.empty()) {
        std::vector<int> ordered;
        ordered.reserve(roster.size());
        for (const std::string& name : startOrder) {
            for (int idx : roster) {
                if (state.statsTable[static_cast<size_t>(idx)].name == name
                    && std::find(ordered.begin(), ordered.end(), idx) == ordered.end()) {
                    ordered.push_back(idx);
                    break;
                }
            }
        }
        // Ce que `startOrder` n'a pas nomme reste dans l'ordre du roster : une
        // manche qui aurait perdu un nom ne doit pas perdre un kart.
        for (int idx : roster) {
            if (std::find(ordered.begin(), ordered.end(), idx) == ordered.end()) {
                ordered.push_back(idx);
            }
        }
        roster = std::move(ordered);
    } else {
        // Fisher-Yates avec le RNG du moteur, jamais le hasard global : c'est ce
        // qui rend une course rejouable a graine egale.
        for (size_t i = roster.size(); i > 1; i--) {
            const size_t j = static_cast<size_t>(rng.next() * static_cast<double>(i));
            std::swap(roster[i - 1], roster[j < i ? j : i - 1]);
        }
    }

    const config::GridCfg& grid = cfg.race.grid;
    const int lanes = static_cast<int>(grid.lanes.size());
    if (lanes <= 0) {
        throw std::runtime_error("createWorldState : la grille n'a aucune colonne");
    }

    const double roadHeight = cfg.road.maxY - cfg.road.minY;
    const int rosterSize = static_cast<int>(roster.size());

    // Moins de personnages actives que de places : la course se fait avec eux,
    // comme en JS. Seul `--karts` (developpement) recycle au-dela du roster.
    const int kartCount = cfg.kartCountForced
        ? cfg.kartCount
        : std::min(cfg.kartCount, rosterSize);

    state.karts.reserve(static_cast<size_t>(kartCount));

    for (int index = 0; index < kartCount; index++) {
        // La grille se DEDUIT du nombre de karts, elle ne le fixe pas : la
        // config decrit une loi de grille (deux colonnes, un pas entre rangs),
        // pas un nombre de places. C'est ce qui permet `--karts=N` entre 1 et 12
        // sans toucher a config/ (plan §3).
        const int row = index / lanes;
        const int col = index % lanes;

        const double gapToLine = grid.backOffset
            + row * grid.rowGap
            + col * grid.colStagger;

        double worldX = cfg.world.finishLineX - gapToLine;
        if (worldX < 0) worldX += cfg.world.width;

        // La diagonale de grille s'accentue avec le rang. Au-dela de 4 rangs —
        // donc au-dela de 8 karts, ce que `--karts` autorise en developpement —
        // elle finirait par poser le dernier kart SUR le bord, dans la bande de
        // frottement : il paierait le bord des le depart. Le pas est donc
        // reparti sur le nombre de rangs REELS plutot que fixe.
        //
        // A 8 karts (4 rangs), le diviseur vaut 4 et la grille est identique a
        // celle du JS, au flottant pres.
        const int rows = (kartCount + lanes - 1) / lanes;
        const double slope = rows > 1
            ? grid.laneSlope * 3.0 / static_cast<double>(rows - 1)
            : 0.0;

        const double depth = grid.lanes[col] + row * slope;
        const double verticalPos = clamp(cfg.road.minY + roadHeight * depth,
                                         cfg.road.minY, cfg.road.maxY);

        // Au-dela du roster, les personnages se RECYCLENT : deux karts peuvent
        // jouer le meme, et partagent alors la meme entree de stats — rien a
        // dupliquer, `stats` est un pointeur (plan §3).
        const int rosterIndex = roster[static_cast<size_t>(index % rosterSize)];

        Kart kart;
        kart.id = index;
        kart.charName = state.statsTable[rosterIndex].name;
        kart.stats = &state.statsTable[rosterIndex];
        if (rosterIndex < static_cast<int>(cfg.bodiesByCharacter.size())) {
            kart.body = cfg.bodiesByCharacter[rosterIndex];
        }

        kart.worldX = worldX;
        kart.yPercent = verticalPos;
        kart.totalDistance = 0;

        // La distance a couvrir, et la raison pour laquelle le classement se
        // fait en distance RESTANTE : deux karts partis de rangs differents
        // n'ont pas la meme a parcourir.
        kart.finishDistance = cfg.race.laps * cfg.world.width + gapToLine;

        kart.state = KartState::Grid;
        kart.rank = index + 1;

        // La profondeur visee part de celle de la grille : sans ca, tout le
        // monde plongerait vers le centre au feu vert.
        kart.laneY = verticalPos;
        kart.nextWanderAt = now + rng.range(cfg.wander.intervalMin, cfg.wander.intervalMax);

        // L'elan initial, tire comme en JS.
        kart.momentumTarget = rng.range(
            cfg.speeds.momentumFloorBase
                + cfg.speeds.momentumFloorWeightGain * kart.stats->normWeight,
            1.0);
        kart.nextMomentumChange = now + rng.range(cfg.speeds.momentumDriftMin,
                                                  cfg.speeds.momentumDriftMax);

        state.karts.push_back(std::move(kart));
    }

    // Le decompte : c'est lui qui ouvre la course, pas le premier tick.
    state.countdownMs = cfg.race.countdownHoldMs + 2 * cfg.race.lightIntervalMs;
    state.startAt = now + state.countdownMs;
    state.phase = Phase::Countdown;

    // ── Le decor, tel que le CIRCUIT le pose ────────────────────────────────
    //
    // `apply_track` a deja traduit les cellules du dessin en px de monde et en
    // profondeur : ici on ne fait que recopier. Un monde sans circuit charge
    // n'a ni boite ni tuyau — et `create_world_state` ne s'en plaint pas, c'est
    // au service de refuser de demarrer sur un dossier tracks/ vide.
    for (const config::Placed& box : cfg.world.itemBoxes) {
        ItemBox b;
        b.worldX = box.x;
        b.y = box.y;
        b.active = true;
        state.itemBoxes.push_back(b);
    }

    for (const config::Placed& pipe : cfg.world.pipes) {
        Pipe p;
        p.worldX = pipe.x;
        p.y = pipe.y;
        p.kind = pipe.red ? 1 : 0;
        state.pipes.push_back(p);
    }

    // La camera se gare face a la ligne pour le depart, pas a l'origine du
    // monde : sinon la premiere image montre une piste vide.
    const double park = cfg.world.finishLineX + cfg.race.parkStartOffset;
    state.cameraX = park < 0 ? park + cfg.world.width : park;
    state.bgCameraX = state.cameraX;

    return state;
}

} // namespace engine
